README.md
# Recoll

Recoll is a desktop full-text search tool. It finds keywords inside
documents as well as file names. 

* Versions are available for Linux and MS Windows.
* A WEB front-end with preview and download features can replace or
  supplement the GUI for remote use. 
* It can search most document formats. You may need external applications
  for text extraction. 
* It can reach any storage place: files, archive members, email
  attachments, transparently handling decompression. 
* One click will open the document inside a native editor or display an
  even quicker text preview. 
* The software is free, open source, and licensed under the GPL.

For more detail, see the [features page on the web site](http://www.recoll.org/features.html) or the [online documentation](http://www.recoll.org/doc.html).

Most distributions feature prebuilt packages for Recoll, but if you need them, [the instructions for building and
installing are here](https://www.lesbonscomptes.com/recoll/usermanual/usermanual.html#RCL.INSTALL.BUILDING).


src/ChangeLog
2009-10-19 12:00  dockes

	* ChangeLog: release 1.12.2 (pre?)

2009-10-19 10:51  dockes

	* qtgui/uiprefs.ui: move the use-desktop-preference checkbox close
	  to the choose editors button

2009-10-19 07:25  dockes

	* rcldb/rcldb.cpp, rcldb/rclquery.cpp: catch xapian exceptions in 2
	  more places.

2009-10-18 07:57  dockes

	* qtgui/reslist.cpp: reslist: rightclick popup would not work
	  inside table

2009-10-17 06:35  dockes

	* Makefile.in, index/recollindex.cpp, qtgui/main.cpp,
	  qtgui/rclmain_w.cpp, rcldb/rcldb.cpp, rcldb/rcldb.h: rclversion.h
	  must not include xapian.h. Replace with Rcl::version_string()

2009-10-17 06:34  dockes

	* ChangeLog:

2009-10-15 12:28  dockes

	* index/rclmonprc.cpp, utils/Makefile, utils/conftree.cpp,
	  utils/execmd.cpp: only call x11IsAlive() from main thread

2009-10-14 14:41  dockes

	* sampleconf/recoll.conf.in:

2009-10-14 14:41  dockes

	* internfile/mh_exec.cpp: no timeout if filtermaxseconds<0

2009-10-13 12:17  dockes

	* internfile/mh_exec.cpp: typo in exec timeout

2009-10-13 11:52  dockes

	* sampleconf/recoll.conf.in: add loop.ps to skippedNames

2009-10-12 17:41  dockes

	* internfile/mh_exec.cpp: configurable timeout on script execution

2009-10-12 16:52  dockes

	* index/recollindex.cpp, internfile/mh_exec.cpp, utils/execmd.cpp:
	  Handle the case of runaway filter processes (ie: loop.ps):
	  - Modify ExecCmd to use a different process group for the son
	  process
	  and use killpg for termination (ensure that all children get
	  killed,
	  not just the top one)
	  - Modify ExecCmd to test for term request on a select() error
	  (avoid calling
	  waitpid() without prior killing in this case).
	  - Modify mh_exec to checkCancel() in the heartbeat/data arrival
	  routine
	  - Modify recollindex to call setCancel when catching a signal

2009-10-12 16:26  dockes

	* common/rclinit.cpp: also block USR1 USR2

2009-10-12 16:26  dockes

	* VERSION:

2009-10-12 11:29  dockes

	* common/textsplit.cpp: merged test driver from 1.13 (no change to
	  main code

2009-08-13 06:36  dockes

	* index/indexer.cpp, internfile/internfile.cpp,
	  internfile/internfile.h, internfile/mimehandler.cpp,
	  internfile/mimehandler.h: xattrs: make them work with non-text
	  files. Use ctime for up to date checks

2009-08-13 06:21  dockes

	* qtgui/rclmain_w.cpp: Make sure db is open at all times (caused
	  problems when sorting query started from the command line)

2009-08-13 06:19  dockes

	* sampleconf/recoll.conf.in: add indexedmimetypes to sample file

2009-07-02 06:18  dockes

	* qtgui/main.cpp, rcldb/searchdata.h, utils/mimeparse.cpp,
	  utils/mimeparse.h: explicitely->explicitly

2009-06-26 09:25  dockes

	* filters/rclman, sampleconf/mimemap: improve man page handling

2009-06-23 06:00  dockes

	* README:

2009-06-23 05:52  dockes

	* ChangeLog, VERSION: version 1.12.1

2009-06-22 08:59  dockes

	* filters/rclman: use groff html output!

2009-06-01 06:32  dockes

	* qtgui/main.cpp: fixed bug in handling remaining arguments as
	  question pieces

2009-05-29 06:28  dockes

	* bincimapmime/convert.h, utils/base64.cpp: change strchr() return
	  parameter to const for new libc

2009-05-25 09:00  dockes

	* filters/rcllyx: bug report from d.prost: spaces and accents in
	  lyx file names

2009-05-04 08:06  dockes

	* qtgui/guiutils.cpp, qtgui/preview_w.h, qtgui/spell_w.cpp: gcc44

2009-04-27 10:42  dockes

	* utils/refcntr.h: MFC

2009-04-27 10:39  dockes

	* utils/md5.cpp: MFC

2009-04-27 10:38  dockes

	* utils/readfile.cpp: MFC

2009-04-27 10:38  dockes

	* utils/refcntr.h: MFC

2009-04-27 10:06  dockes

	* qtgui/i18n, qtgui/i18n/recoll_ru.ts, qtgui/i18n/recoll_uk.ts:
	  brought trunk changes into 1.12maint

2009-04-27 09:23  dockes

	* python/recoll/setup.py: pathhash->fileudi

2009-04-27 09:20  dockes

	* Maintenance branch for 1.12

2009-01-30 13:28  dockes

	* Release 1.12.0 tagged

2009-01-30 11:42  dockes

	* 1.12 manual

2009-01-30 10:22  dockes

	* updated message files, translated french

2009-01-29 14:24  dockes

	* simplified rcltext. No need for awk and no assumptions on charset

2009-01-29 11:27  dockes

	* 

2009-01-29 11:04  dockes

	* 1.12.0 une

2009-01-29 10:08  dockes

	* have ssearch install the lang help section when needed

2009-01-28 17:41  dockes

	* erase history would crash with empty reslist docsource

2009-01-28 17:21  dockes

	* fixed status bar messages (were cleared by periodic100 every
	  100ms)

2009-01-28 17:05  dockes

	* F1 context-enhanced help

2009-01-28 16:56  dockes

	* F1 context-enhanced help

2009-01-28 16:56  dockes

	* F1 context-enhanced help

2009-01-28 14:58  dockes

	* dont encode urls used for starting help browser

2009-01-28 14:22  dockes

	* add xapian version to version string

2009-01-28 08:45  dockes

	* avoid setting 0 sizes

2009-01-28 08:40  dockes

	* allow setting QMAKE in the environment

2009-01-27 18:12  dockes

	* 

2009-01-27 10:25  dockes

	* Emit a_b intermediary span when splitting a_b.c

2009-01-26 18:30  dockes

	* modified the time at which we unaccent so that we can do the
	  Capitalized->nostemming test on single words (this had been
	  broken by the change of noac/split order done earlier to get
	  japanese to work)

2009-01-26 13:29  dockes

	* add overloaded neutchars with different parameters

2009-01-26 13:27  dockes

	* tested and decided against cacheing iconv_open

2009-01-23 15:56  dockes

	* temp ckpt: need to test on real unix

2009-01-23 11:07  dockes

	* avoid name duplication

2009-01-23 11:03  dockes

	* one button for choosing native editors

2009-01-23 10:38  dockes

	* simplified javascrip: no ie here!

2009-01-23 09:41  dockes

	* toLocal8Bit->local8bit

2009-01-23 09:30  dockes

	* use normal text/html ext app for viewing help

2009-01-23 09:27  dockes

	* accept additional path argument to execmd::which

2009-01-22 14:25  dockes

	* allow toggle show text/fields in preview

2009-01-21 16:42  dockes

	* added saveToFile menu entry to reslist


src/rcldb/rcldb.h
/*
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the
 *   Free Software Foundation, Inc.,
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
#ifndef _DB_H_INCLUDED_
#define _DB_H_INCLUDED_
/* @(#$Id: rcldb.h,v 1.65 2008-12-19 09:55:36 dockes Exp $  (C) 2004 J.F.Dockes */

#include <string>
#include <list>
#include <vector>

#include "refcntr.h"
#include "rcldoc.h"
#include "stoplist.h"

#ifndef NO_NAMESPACES
using std::string;
using std::list;
using std::vector;
#endif

// rcldb defines an interface for a 'real' text database. The current 
// implementation uses xapian only, and xapian-related code is in rcldb.cpp
// If support was added for other backend, the xapian code would be moved in 
// rclxapian.cpp, another file would be created for the new backend, and the
// configuration/compile/link code would be adjusted to allow choosing. There 
// is no plan for supporting multiple different backends.
// 
// In no case does this try to implement a useful virtualized text-db interface
// The main goal is simplicity and good matching to usage inside the recoll
// user interface. In other words, this is not exhaustive or well-designed or 
// reusable.
//
// Unique Document Identifier: unically identifies a document in its
// source storage (file system or other). Used for up to date checks
// etc. "udi". Our user is responsible for making sure it's not too
// big, cause it's stored as a Xapian term (< 150 bytes would be
// reasonable)

class RclConfig;

#ifndef NO_NAMESPACES
namespace Rcl {
#endif

class SearchData;
class TermIter;
class Query;

/** Used for returning result lists for index terms matching some criteria */
class TermMatchEntry {
public:
    TermMatchEntry() : wcf(0) {}
    TermMatchEntry(const string&t, int f) : term(t), wcf(f) {}
    TermMatchEntry(const string&t) : term(t), wcf(0) {}
    bool operator==(const TermMatchEntry &o) { return term == o.term;}
    bool operator<(const TermMatchEntry &o) { return term < o.term;}
    string term;
    int    wcf; // Within collection frequency
};

/**
 * Wrapper class for the native database.
 */
class Db {
 public:
    // A place for things we don't want visible here.
    class Native;
    friend class Native;

    /* General stuff (valid for query or update) ****************************/
    Db(RclConfig *cfp);
    ~Db();

    enum OpenMode {DbRO, DbUpd, DbTrunc};
    bool open(OpenMode mode, bool keep_updated = false);
    bool close();
    bool isopen();

    /** Get explanation about last error */
    string getReason() const {return m_reason;}

    /** List possible stemmer names */
    static list<string> getStemmerNames();

    /** List existing stemming databases */
    std::list<std::string> getStemLangs();

    /* The next two, only for searchdata, should be somehow hidden */
    /* Return list of configured stop words */
    const StopList& getStopList() const {return m_stops;}
    /* Field name to prefix translation (ie: author -> 'A') */
    bool fieldToPrefix(const string& fldname, string &pfx);

    /* Update-related methods ******************************************/

    /** Test if the db entry for the given udi is up to date (by
     * comparing the input and stored sigs). 
     * Side-effect: set the existence flag for the file document 
     * and all subdocs if any (for later use by 'purge()') 
     */
    bool needUpdate(const string &udi, const string& sig);

    /** Add or update document. The Doc class should have been filled as much as
      * possible depending on the document type. parent_udi is only
      * use for subdocs, else set it to empty */
    bool addOrUpdate(const string &udi, const string &parent_udi, 
		     const Doc &doc);

    /** Delete document(s) for given UDI, including subdocs */
    bool purgeFile(const string &udi);

    /** Remove documents that no longer exist in the file system. This
     * depends on the update map, which is built during
     * indexation (needUpdate()). 
     *
     * This should only be called after a full walk of
     * the file system, else the update map will not be complete, and
     * many documents will be deleted that shouldn't, which is why this
     * has to be called externally, rcldb can't know if the indexing
     * pass was complete or partial.
     */
    bool purge();

    /** Create stem expansion database for given language. */
    bool createStemDb(const string &lang);
    /** Delete stem expansion database for given language. */
    bool deleteStemDb(const string &lang);

    /* Query-related methods ************************************/

    /** Return total docs in db */
    int  docCnt(); 

    /** Add extra database for querying */
    bool addQueryDb(const string &dir);
    /** Remove extra database. if dir == "", remove all. */
    bool rmQueryDb(const string &dir);
    /** Tell if directory seems to hold xapian db */
    static bool testDbDir(const string &dir);

    /** Return a list of index terms that match the input string
     * Expansion is performed either with either wildcard or regexp processing
     * Stem expansion is performed if lang is not empty */
    enum MatchType {ET_WILD, ET_REGEXP, ET_STEM};
    bool termMatch(MatchType typ, const string &lang, const string &s, 
		   list<TermMatchEntry>& result, int max = -1, 
		   const string& field = "");

    /** Special filename wildcard to XSFN terms expansion.
	internal/searchdata use only */
    bool filenameWildExp(const string& exp, list<string>& names);

    /** Set parameters for synthetic abstract generation */
    void setAbstractParams(int idxTrunc, int synthLen, int syntCtxLen);

    /** Build synthetic abstract for document, extracting chunks relevant for
     * the input query. This uses index data only (no access to the file) */
    bool makeDocAbstract(Doc &doc, Query *query, string& abstract);

    /** Get document for given udi
     *
     * Used by the 'history' feature (and nothing else?) 
     */
    bool getDoc(const string &udi, Doc &doc);

    /* The following are mainly for the aspell module */
    /** Whole term list walking. */
    TermIter *termWalkOpen();
    bool termWalkNext(TermIter *, string &term);
    void termWalkClose(TermIter *);
    /** Test term existence */
    bool termExists(const string& term);
    /** Test if terms stem to different roots. */
    bool stemDiffers(const string& lang, const string& term, 
		     const string& base);

    RclConfig *getConf() {return m_config;}

    /* This has to be public for access by embedded Query::Native */
    Native *m_ndb; 

private:
    // Internal form of close, can be called during destruction
    bool i_close(bool final);

    RclConfig *m_config;
    string     m_reason; // Error explanation

    /* Parameters cached out of the configuration files */
    // This is how long an abstract we keep or build from beginning of
    // text when indexing. It only has an influence on the size of the
    // db as we are free to shorten it again when displaying
    int          m_idxAbsTruncLen;
    // This is the size of the abstract that we synthetize out of query
    // term contexts at *query time*
    int          m_synthAbsLen;
    // This is how many words (context size) we keep around query terms
    // when building the abstract
    int          m_synthAbsWordCtxLen;
    // Flush threshold. Megabytes of text indexed before we flush.
    int          m_flushMb;
    // Text bytes indexed since beginning
    long long    m_curtxtsz;
    // Text bytes at last flush
    long long    m_flushtxtsz;
    // Text bytes at last fsoccup check
    long long    m_occtxtsz;
    // Maximum file system occupation percentage
    int          m_maxFsOccupPc;

    // Database directory
    string       m_basedir;

    // List of directories for additional databases to query
    list<string> m_extraDbs;

    OpenMode m_mode;

    vector<bool> updated;

    StopList m_stops;

    bool reOpen(); // Close/open, same mode/opts
    bool stemExpand(const string &lang, const string &s, 
		    list<TermMatchEntry>& result, int max = -1);

    /* Copyconst and assignemt private and forbidden */
    Db(const Db &) {}
    Db& operator=(const Db &) {return *this;};
};

// This has to go somewhere, and as it needs the Xapian version, this is
// the most reasonable place.
string version_string();

#ifndef NO_NAMESPACES
}
#endif // NO_NAMESPACES


#endif /* _DB_H_INCLUDED_ */
